openide.awt/src/org/openide/awt/GeneralAction.java

```diff
@@ -74,24 +80,77 @@ public static ContextAwareAction callback(Map map) {
         Lookup context, 
         Class<T> dataType
     ) {
-        return new ContextAction<T>(perf, selectionType, context, dataType, false);
+        return new ContextAction<T>(perf, selectionType, context, dataType, false, null);
     }
     
     public static ContextAwareAction context(Map map) {
         Class<?> dataType = readClass(map.get("type")); // NOI18N
-        return new DelegateAction(map, _context(map, dataType, Utilities.actionsGlobalContext()));
+        ContextAwareAction ca = _context(map, dataType, Utilities.actionsGlobalContext());
+        // autodetect on/off actions
+        if (ca.getValue(Action.SELECTED_KEY) != null) {
+            return new StateDelegateAction(map, ca);
+        } else {
+            return new DelegateAction(map, ca);
+        }
     }
+    
     public static Action bindContext(Map map, Lookup context) {
         Class<?> dataType = readClass(map.get("type")); // NOI18N
         return new BaseDelAction(map, _context(map, dataType, context));
     }
+    
     private static <T> ContextAwareAction _context(Map map, Class<T> dataType, Lookup context) {
         ContextSelection sel = readSelection(map.get("selectionType")); // NOI18N
         Performer<T> perf = new Performer<T>(map);
         boolean survive = Boolean.TRUE.equals(map.get("surviveFocusChange")); // NOI18N
-        return new ContextAction<T>(
-            perf, sel, context, dataType, survive
-        );
+        StatefulMonitor enableMonitor = null;
+        StatefulMonitor checkMonitor = null;
+        Class enableType = tryReadClass(map.get("enableOnType"));
+        if (enableType == null) {
+            enableType = dataType;
+        }
+        Object del = map.get("enableOnActionProperty");
+        Object o = map.get("enableOnProperty"); // NOI18N
+        
+        if (o instanceof String || (o == null && (del instanceof String))) {
+            enableMonitor = new PropertyMonitor(enableType, (String)o, "enableOn", map);
+        }
+        o = map.get("checkedOnProperty"); // NOI18N
+        if (o instanceof String) {
+            Class c = tryReadClass(map.get("checkedOnType")); // NOI18N
+            if (c != null) {
+                checkMonitor = new PropertyMonitor(c, (String)o, "checkedOn", map);
+            }
+        }
+        // special case to hook on existing action instances
+        if (map.containsKey("_delegateReady")) { // NOI18N
```

Comment:
Nasty. It is not common to prefix properties with `_`. Possibly document as layer based API with lower stability. Probably give it some less scary name.

Reply:
It is nasty. In fact it is probably even not an API (at least not one I'd like to maintain), because it serves just as a private channel between `Actions.context(Class, boolean, boolean, delegate, ...)` and the implementation. I was lazy to change method signatures. Will remove/do other way.